### PR TITLE
Use the view owner's identity even if the current user is the owner

### PR DESCRIPTION
**Pull request by ksobolew**
## Description

On the surface it makes sense that if the current session's user is the owner, then it is not necessary to substitute the identity when checking permissions for views. The issue is, though, that the current identity and the substituted one may have a different set of enabled roles. If that happens, then the access permissions may be different (usually lessened) when the view is queried by the owner.

Depends on #14130, because with this change it's now possible to trigger that bug.

> Is this change a fix, improvement, new feature, refactoring, or other?

Fix

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

Core engine (access control)

> How would you describe this change to a non-technical end user or system administrator?

When a view owner queries a view with SECURITY DEFINER, the owner's granted roles are used instead of currently enabled roles.

## Documentation

(x) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
(x) Release notes entries required with the following suggested text:

```markdown
# Security
* When a view owner queries a view with SECURITY DEFINER, the owner's granted roles are used instead of currently enabled roles.
```

### CODE REVIEW

File: testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java [hunk 1]
```diff
@@ -195,10 +197,12 @@ public void testViewColumnAccessControl()
                 privilege(viewOwnerSession.getUser(), "orders", CREATE_VIEW_WITH_SELECT_COLUMNS));
 
         // verify the view owner can select from the view even without special view creation privileges
+/* TODO: this is no longer true! the owner has no special privileges anymore
```
Comment (kokosing) [inline testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java:200]: How that happened?

File: core/trino-main/src/main/java/io/trino/server/testing/TestingSystemSecurityMetadata.java [hunk 2]
```diff
@@ -221,7 +238,29 @@ public void setTableOwner(Session session, CatalogSchemaTableName table, TrinoPr
     @Override
     public Optional<Identity> getViewRunAsIdentity(Session session, CatalogSchemaTableName viewName)
     {
-        return Optional.empty();
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/server/testing/TestingSystemSecurityMetadata.java:224]: Why do you need to change this?

File: testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java (hunk 1 above)
Comment (ksobolew) [reply testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java:200]: Previously, if the current session's user was the view's owner, it fell into this branch in `StatementAnalyzer`:

```java
                else {
                    identity = session.getIdentity();
                    viewAccessControl = accessControl;
                }
```

Which means that the view owner was treated as if the view was `SECURITY INVOKER`. Which goes slightly further than "the current session's enabled roles are used".

Honestly, I wasn't expecting this. Now the question is if this is the expected semantics which should be preserved (entirely or in part).

File: core/trino-main/src/main/java/io/trino/server/testing/TestingSystemSecurityMetadata.java (hunk 2 above)
Comment (ksobolew) [reply core/trino-main/src/main/java/io/trino/server/testing/TestingSystemSecurityMetadata.java:224]: This method is used to determine the actual view owner's identity, in context of the currently enabled security system. Returning `empty()` from here will make the analyzer create a bare-bones identity with no enabled roles.

File: testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java (hunk 1 above)
Comment (kokosing) [reply testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java:200]: It is expected semantics. View owner should have the access to its view.

File: core/trino-main/src/main/java/io/trino/server/testing/TestingSystemSecurityMetadata.java (hunk 2 above)
Comment (kokosing) [reply core/trino-main/src/main/java/io/trino/server/testing/TestingSystemSecurityMetadata.java:224]: It is testing metadata, can we simply inject an owner?

File: testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java (hunk 1 above)
Comment (kokosing) [reply testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java:200]: What about https://gist.github.com/kokosing/3b0dabfda2d85f08c03c7d669dd06c50? I am not sure if we should repeat that for connector access controls.

Comment (ksobolew) [reply testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java:200]: I am not sure about `.withEnabledRoles(metadata.listRoles(session, Optional.empty()))`, but otherwise seems like the way to go

Comment (ksobolew) [reply testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java:200]: > I am not sure about...

By that I mean hat it should be delegated to the system access control instead of assuming that surely we want all granted roles to be enabled.

File: core/trino-main/src/main/java/io/trino/server/testing/TestingSystemSecurityMetadata.java (hunk 2 above)
Comment (ksobolew) [reply core/trino-main/src/main/java/io/trino/server/testing/TestingSystemSecurityMetadata.java:224]: It's kind of simpler to just implement this method :)

**Comment by findepi** [comment]
> ```gfm
> # Security
> * When checking permissions for a view with `SECURITY DEFINER`, the permissions checks are now more consistent when the invoker is the owner.
> ```



What does "more consistent" means?
What's the end-user implication of the change?

**Comment by ksobolew** [comment]
> What does "more consistent" means?
> What's the end-user implication of the change?

Man, it's so hard to summarize a subtle change like this...

"When a view owner queries a view with `SECURITY DEFINER`, the owner's granted roles are correctly taken into account."

Is it better? :)

**Comment by ksobolew** [comment]
I dropped the entire commit "Install TestingSystemSecurityMetadata in TestingTrinoServer", it's way too much for my needs here.

File: core/trino-spi/src/main/java/io/trino/spi/security/Identity.java [hunk 3]
```diff
@@ -130,13 +130,15 @@ public boolean equals(Object o)
             return false;
         }
         Identity identity = (Identity) o;
-        return Objects.equals(user, identity.user);
+        return Objects.equals(user, identity.user)
```
Comment (kokosing) [inline core/trino-spi/src/main/java/io/trino/spi/security/Identity.java:133]: I would extract this commit to separate pull request and also I would love to add a test to `TestAccessControl` to expose the issue it may cause.

I think we could consider skipping groups as they are not used access control.

I mean separate pull request is needed as this changes the semantic so it is dense.

File: core/trino-main/src/main/java/io/trino/server/testing/TestingSystemSecurityMetadata.java (hunk 2 above)
Comment (kokosing) [reply core/trino-main/src/main/java/io/trino/server/testing/TestingSystemSecurityMetadata.java:224]: The current implementation it is not necessarly the correct way of getting the owner. It may differ between access controls. Injecting is always correct for tests as leave it for a test and it is possible to test different scenarios.

File: testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java [hunk 4]
```diff
@@ -260,6 +276,30 @@ public void testViewColumnAccessControl()
                 "Cannot select from columns \\[.*] in table .*.orders.*",
                 privilege(getSession().getUser(), "orders", SELECT_COLUMN));
 
+        // verify that role permissions are taken for the view owner's role grants
```
Comment (kokosing) [inline testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java:279]: maybe separate test method?

File: testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java [hunk 5]
```diff
@@ -184,11 +196,15 @@ public void testAccessControl()
     @Test
     public void testViewColumnAccessControl()
     {
+        TestingSystemSecurityMetadata systemSecurityMetadata = (TestingSystemSecurityMetadata) getDistributedQueryRunner().getCoordinator().getInstance(Key.get(SystemSecurityMetadata.class));
```
Comment (kokosing) [inline testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java:199]: maybe add a method to get `TestingSystemSecurityMetadata` to `QueryRunner`

File: core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java [hunk 6]
```diff
@@ -3981,15 +3981,22 @@ private RelationType analyzeView(Query query, QualifiedObjectName name, Optional
...
+                    // don't restrict owner's access
+                    viewAccessControl = accessControl;
                 }
                 else {
-                    identity = session.getIdentity();
-                    viewAccessControl = accessControl;
+                    identity = Identity.from(owner.get())
+                            .withGroups(groupProvider.getGroups(owner.get().getUser()))
+                            .build();
+                    viewAccessControl = new ViewAccessControl(accessControl, session.getIdentity());
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:3999]:
```
                if (owner.isEmpty()) {
                    identity = session.getIdentity();
                    viewAccessControl = accessControl;
                }
                else {
                       identity = Identity.from(owner.get())
                            .withGroups(groupProvider.getGroups(owner.get().getUser()))
                            .build();
                    if (owner.get().getUser().equals(session.getIdentity().getUser())) {
                        // don't restrict owner's access
                        viewAccessControl = accessControl;
                      }
                      else {
                        viewAccessControl = new ViewAccessControl(accessControl, session.getIdentity());
                    }
                }
```

File: core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java [hunk 7]
```diff
@@ -3981,15 +3981,22 @@ private RelationType analyzeView(Query query, QualifiedObjectName name, Optional
                 // run view as view owner if set; otherwise, run as session user
                 Identity identity;
                 AccessControl viewAccessControl;
-                if (owner.isPresent() && !owner.get().getUser().equals(session.getIdentity().getUser())) {
+                if (owner.isEmpty()) {
+                    identity = session.getIdentity();
+                    viewAccessControl = accessControl;
+                }
+                else if (owner.get().getUser().equals(session.getIdentity().getUser())) {
                     identity = Identity.from(owner.get())
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:3989]: why we don't need to get a list of all roles?

File: testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java [hunk 8]
```diff
@@ -260,6 +276,30 @@ public void testViewColumnAccessControl()
                 "Cannot select from columns \\[.*] in table .*.orders.*",
                 privilege(getSession().getUser(), "orders", SELECT_COLUMN));
 
+        // verify that role permissions are taken for the view owner's role grants
+        // even if executed by the owner with different roles enabled
+        executeExclusively(() -> {
+            systemSecurityMetadata.grantRoles(getSession(), Set.of("deniedrole"), Set.of(new TrinoPrincipal(USER, "test_view_access_owner")), false, Optional.empty());
+            try {
+                getQueryRunner().getAccessControl().denyIdentityTable((identity, table) -> !identity.getEnabledRoles().contains("deniedrole") || !"orders".equals(table));
```
Comment (kokosing) [inline testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java:284]: I would implement this differently without using deny. I view owner could have two roles where one has access to base table, then view owner with session when that role is not set would try to query the view and we would need to see it is passing while previously it was failing.

File: core/trino-spi/src/main/java/io/trino/spi/security/Identity.java (hunk 3 above)
Comment (ksobolew) [reply core/trino-spi/src/main/java/io/trino/spi/security/Identity.java:133]: > I would extract this commit to separate pull request and also I would love to add a test to `TestAccessControl` to expose the issue it may cause.

I'm not sure if this can be triggered currently. I know of only one place (so far) where this may cause issues and it would require somehow requiring two access checks with the same use but with different sets of enabled roles. This can only happen when views are involved but, before the change in this PR, whenever a view changes the identity, it always changes the user.

File: testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java (hunk 5 above)
Comment (ksobolew) [reply testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java:199]: Are we "allowed" to add such methods for bindings which are not guaranteed to be to the expected class? I mean `SystemSecurityMetadata` is bound to `TestingSystemSecurityMetadata` in only a couple of tests.

File: core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java (hunk 7 above)
Comment (ksobolew) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:3989]: The list of roles is actually provided by the access control via `io.trino.metadata.SystemSecurityMetadata#getViewRunAsIdentity` and it appears here as the `owner` `Identity`. If we get all the roles here, we kind of enforce this semantics on all access control implementations.

File: core/trino-spi/src/main/java/io/trino/spi/security/Identity.java (hunk 3 above)
Comment (ksobolew) [reply core/trino-spi/src/main/java/io/trino/spi/security/Identity.java:133]: I extracted #14130, but in the meantime I discovered some more questions :) (They're in a comment in that PR)

File: testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java (hunk 8 above)
Comment (ksobolew) [reply testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java:284]: I would like to do it that way too, but `TestingAccessControlManager` can only deny stuff. Unless there is a system access control installed where we can inject policies...

**Comment by kokosing** [comment]
Let's wait on https://github.com/trinodb/trino/pull/14130 before moving forward with this.

**Comment by kokosing** [comment]
Replaced with https://github.com/trinodb/trino/pull/15111
